libvast/vast/system/spawn_source.hpp

```diff
@@ -14,50 +14,59 @@
+#include "vast/detail/make_io_stream.hpp"
+#include "vast/detail/unbox_var.hpp"
 #include "vast/fwd.hpp"
+#include "vast/logger.hpp"
+#include "vast/system/node.hpp"
+#include "vast/system/source.hpp"
+#include "vast/system/spawn_arguments.hpp"
 
 namespace vast::system {
 
-/// Tries to spawn a new SOURCE for the CSV format.
+/// Tries to spawn a new SOURCE for the specified format.
+/// @tparam Reader the format-specific reader
 /// @param self Points to the parent actor.
 /// @param args Configures the new actor.
 /// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_csv_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the JSON format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_json_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the PCAP format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_pcap_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the Suricata format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_suricata_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the Syslog format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_syslog_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for generating random data.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_test_source(node_actor* self, spawn_arguments& args);
-
-/// Tries to spawn a new SOURCE for the Zeek format.
-/// @param self Points to the parent actor.
-/// @param args Configures the new actor.
-/// @returns a handle to the spawned actor on success, an error otherwise
-maybe_actor spawn_zeek_source(node_actor* self, spawn_arguments& args);
+template <class Reader, class Defaults = typename Reader::defaults>
+maybe_actor spawn_source(node_actor* self, spawn_arguments& args) {
+  VAST_TRACE(VAST_ARG("node", self), VAST_ARG(args));
+  auto& st = self->state;
+  auto& options = args.inv.options;
+  // Bail out early for bogus invocations.
+  if (caf::get_or(options, "system.node", false))
+    return make_error(ec::parse_error, "cannot start a local node");
+  VAST_UNBOX_VAR(in, detail::make_input_stream<Defaults>(args.inv.options));
+  VAST_UNBOX_VAR(sch, read_schema(args));
```

Comment:
I'm not a fan of these macros, but we can keep them if you like 'em. To me, the make it more opaque what's actually going on here.

Reply:
Let's make that a separate story—we should just remove them everywhere. I did not really touch this code, I just moved it to a separate place.